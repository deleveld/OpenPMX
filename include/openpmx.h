/*
 * This file is part of OpenPMX (https://github.com/deleveld/openpmx).
 * Copyright (c) 2022 Douglas Eleveld.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, version 3.
 *
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */

#ifndef OPENPMX_H
#define OPENPMX_H

#include <stdbool.h>

#ifdef __cplusplus
extern "C" {
#endif

/*---------------------------------------------------------------------*/
#define OPENPMX_VERSION_MAJOR			0
#define OPENPMX_VERSION_MINOR			0
#define OPENPMX_VERSION_RELEASE			1

#define OPENPMX_THETA_MAX				64
#define OPENPMX_OMEGABLOCKSIZE_MAX		64
#define OPENPMX_OMEGABLOCK_MAX			5
#define OPENPMX_OMEGA_MAX				16
#define OPENPMX_SIGMA_MAX				8

#define OPENPMX_FIELDS_MAX				64
#define OPENPMX_FIELDNAME_MAX			64

#define OPENPMX_STATE_MAX				32
#define OPENPMX_IMODEL_MAX				64
#define OPENPMX_PREDICTVARS_MAX			64
#define OPENPMX_SIMULINFUSION_MAX		16

/*---------------------------------------------------------------------*/
/* dataconfig */
/*---------------------------------------------------------------------*/
typedef	struct {
	int size;
	struct {
		char name[OPENPMX_FIELDNAME_MAX];
		int offset;
	} field[OPENPMX_FIELDS_MAX];
} STRUCTINFO;

typedef struct RECORD RECORD;
typedef struct {
	RECORD* writeable;
	const RECORD* records;
	const int nrecords;
	const bool _offset1;
	const STRUCTINFO recordfields;
} DATACONFIG;

/*---------------------------------------------------------------------*/
/* advan */
/*---------------------------------------------------------------------*/
typedef struct ADVAN ADVAN;
typedef struct {
	/* opaque pointer into the ADVAN so we can change setting in init
	 * and predict functions */
	ADVAN* advan;
	
	/* info about the record being processed */
	const double statetime;
	const RECORD* const record;
	const double* const state;
} ADVANSTATE;

int pmx_advan_initcount(const ADVANSTATE* const advanstate);
void pmx_advan_amtlag(const ADVANSTATE* const advanstate, const int cmt, const double t);
void pmx_advan_bioaval(const ADVANSTATE* const advanstate, const int cmt, const double f);
void pmx_advan_inittime(const ADVANSTATE* const advanstate, const double t);
void pmx_advan_state_init(const ADVANSTATE* const advanstate, const int cmt, const double v);

typedef	struct {
	const double* const theta;
	const int ntheta;
	const double* const eta;
	const int nomega;
	const double* const sigma;
	const int nsigma;
	const int nstate;
} POPPARAM;

/* must be initialized in IMODEL_INIT */
typedef struct IMODEL IMODEL;
typedef void (*IMODEL_INIT)(IMODEL* const imodel,
							ADVANSTATE* const advanstate,
							const POPPARAM* const popparam);

typedef struct {
	const double* const state;
	const RECORD* const record;
} PREDICTSTATE;

typedef struct PREDICTVARS PREDICTVARS;
typedef	double (*IMODEL_PREDICT)(const IMODEL* const imodel,
								 const PREDICTSTATE* const predictstate,
								 const POPPARAM* const popparam,
								 const double* const err,
								 PREDICTVARS* predparams);

typedef void (*ADVAN_DIFFEQN)(double DADT[],
							  const IMODEL* const imodel,
							  const RECORD* const record,
							  const double* state,
							  const POPPARAM* const popparam,
							  const double T);

/* unified arguments for all advancer functions */
typedef struct ADVANCONFIG ADVANCONFIG;
typedef struct ADVANFUNCS ADVANFUNCS;
typedef	ADVANFUNCS* (*ADVANMETHOD)(const DATACONFIG* const dataconfig, const ADVANCONFIG* const advanconfig);

typedef struct ADVANCONFIG {
	const IMODEL_INIT init;
	const IMODEL_PREDICT predict;
	const int firstonly;
	const int predictall;

	const STRUCTINFO imodelfields;
	const STRUCTINFO predictfields;
	const ADVANMETHOD method;
	const int nstate;
	const ADVAN_DIFFEQN diffeqn;

	/* extra arguments */
	union {
		/* for diffeqn solvers */
		struct {
			const double hstart;
			const double abstol;
			const double reltol;
			const char* steptype;
		} diffeqn;
	} args;
} ADVANCONFIG;

/* advancer methods */
ADVANFUNCS* pmx_advan_pred(const DATACONFIG* const dataconfig, const ADVANCONFIG* const advanconfig);
ADVANFUNCS* pmx_advan_onecomp(const DATACONFIG* const dataconfig, const ADVANCONFIG* const advanconfig);
ADVANFUNCS* pmx_advan_onecomp_depot(const DATACONFIG* const dataconfig, const ADVANCONFIG* const advanconfig);
ADVANFUNCS* pmx_advan_twocomp(const DATACONFIG* const dataconfig, const ADVANCONFIG* const advanconfig);
ADVANFUNCS* pmx_advan_threecomp(const DATACONFIG* const dataconfig, const ADVANCONFIG* const advanconfig);

ADVANFUNCS* pmx_advan_diffeqn_libgsl(const DATACONFIG* const dataconfig, const ADVANCONFIG* const advanconfig);

ADVANFUNCS* pmx_advan_diffeqn_test(const DATACONFIG* const dataconfig, const ADVANCONFIG* const advanconfig);
ADVANFUNCS* pmx_advan_wrapper(const DATACONFIG* const dataconfig, const ADVANCONFIG* const advanconfig);

/*---------------------------------------------------------------------*/
/* basic types */
/*---------------------------------------------------------------------*/
typedef struct {
	double objfn;
	enum {
		OBJFN_INVALID = 0,
		OBJFN_EVALUATE,
		OBJFN_CURRENT,
		OBJFN_FINAL,
	} type;
	int neval;
} PMXRESULT;

/*---------------------------------------------------------------------*/
/* OPENPMX */
/*---------------------------------------------------------------------*/
typedef struct {
	/* options and config */
	const char* filename;
	int nthread;
	bool _offset1;

	/* the problem structure to analyse */
	const DATACONFIG data;
	const ADVANCONFIG advan;

	/* the population paramaters and results of last run */
	struct {
		double lower, value, upper;
		enum {
			THETA_INVALID = 0,
			FIXED = 1,
			ESTIMATE = 2,
		} type;
	} theta[OPENPMX_THETA_MAX];
	struct {
		enum {
			OMEGA_INVALID = 0,
			OMEGA_DIAG,
			OMEGA_BLOCK,
			OMEGA_SAME,
		} type;
		int ndim;
		double values[OPENPMX_OMEGABLOCKSIZE_MAX];
	} omega[OPENPMX_OMEGABLOCK_MAX];
	double sigma[OPENPMX_SIGMA_MAX];

	/* output */
	PMXRESULT result;

	/* internal use */
	struct PMXSTATE* state;
} OPENPMX;

void pmx_cleanup(OPENPMX* openpmx);

OPENPMX pmx_copy(const OPENPMX* const openpmx);
void pmx_copy_popparams(OPENPMX* dest, const OPENPMX* const src);

/*---------------------------------------------------------------------*/
/* prediction */
/*---------------------------------------------------------------------*/
void pmx_predict(OPENPMX* openpmx);
void pmx_predict_pred(OPENPMX* openpmx);
const IMODEL* pmx_imodel(OPENPMX* const pmx);
const PREDICTVARS* pmx_predictvars(OPENPMX* const pmx);

/*---------------------------------------------------------------------*/
/* simulation */
/*---------------------------------------------------------------------*/
typedef struct {
	unsigned long seed;
} SIMCONFIG;

void pmx_resample(OPENPMX* pmx, const SIMCONFIG* const simconfig);
void pmx_simulate(OPENPMX* pmx, const SIMCONFIG* const simconfig);

/*---------------------------------------------------------------------*/
/* evaluation */
/*---------------------------------------------------------------------*/
typedef struct {
	double gradient_step;
	double step_initial;
	double step_refine;
	double step_final;
	int maxeval;

	bool icov_resample;
} STAGE1CONFIG;

void pmx_evaluate(OPENPMX* pmx, STAGE1CONFIG* const evalconfig);

/*---------------------------------------------------------------------*/
/* estimation */
/*---------------------------------------------------------------------*/
typedef struct {
	STAGE1CONFIG stage1;

	double step_initial;
	double step_refine;
	double step_final;
	int maxeval;
	double dobjfn;

	bool details;
	bool verbose;
	bool progress;
} ESTIMCONFIG;

void pmx_estimate(OPENPMX* pmx, ESTIMCONFIG* const estimconfig);
void pmx_fastestimate(OPENPMX* pmx, ESTIMCONFIG* const estimconfig);

/*---------------------------------------------------------------------*/
/* tables */
/*---------------------------------------------------------------------*/
typedef struct {
	const char* filename;
	const char* name;
	bool firstonly;
} TABLECONFIG;

void pmx_table(OPENPMX* pmx, const char* fields, const TABLECONFIG* const tableconfig);

#ifdef __cplusplus
}
#endif

/* OPENPMX_H */
#endif
