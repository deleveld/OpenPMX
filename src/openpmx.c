/* 
 * This file is part of OpenPMX (https://github.com/deleveld/openpmx).
 * Copyright (c) 2022 Douglas Eleveld.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, version 3.
 *
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */

#include <assert.h>
#include <values.h>
#include <unistd.h>

#include "openpmx.h"
#include "print.h"
#include "scatter.h"
#include "utils/c22.h"
#include "advan/advan.h"
#include "openpmx_internal.h"

#include "openpmx_compile_options.h"

STAGE1CONFIG stage1config_default(const STAGE1CONFIG* const stage1)
{
	STAGE1CONFIG ret = { 0 };
	if (stage1)
		ret = *stage1;

	let v = pow(DBL_EPSILON, 1./3.);
	if (ret.gradient_step == 0.)
		ret.gradient_step = 1e-4;
	if (ret.step_initial == 0.)
		ret.step_initial = 0.2;
	if (ret.step_refine == 0.)
		ret.step_refine = 0.1;
	if (ret.step_final == 0.)
		ret.step_final = v;
	if (ret.maxeval == 0)
		ret.maxeval = 1000;

	/* 	default omit_icov_resample = false */
	if (ret.icov_resample_tol == 0)
		ret.icov_resample_tol = DBL_MAX;

	return ret;
}

ESTIMCONFIG estimconfig_default(const ESTIMCONFIG* const estimate)
{
	ESTIMCONFIG ret = { 0 };
	if (estimate)
		ret = *estimate;

	ret.stage1 = stage1config_default(&ret.stage1);

	let v = pow(DBL_EPSILON, 1./3.);
	if (ret.optim.step_initial == 0.)
		ret.optim.step_initial = 0.2;
	if (ret.optim.step_refine == 0.)
		ret.optim.step_refine = 0.05;
	if (ret.optim.step_final == 0.)
		ret.optim.step_final = v * 10.;
	if (ret.optim.maxeval == 0)
		ret.optim.maxeval = 10000;

	if (ret.posthoc.deriv.step == 0.)
		ret.posthoc.deriv.step = ret.optim.step_initial;

	return ret;
}

SIMCONFIG simconfig_default(const SIMCONFIG* const simulate)
{
	SIMCONFIG ret = { 0 };
	if (simulate)
		ret = *simulate;

	if (ret.seed == 0)
		ret.seed = 200501041406; /* Birthdays of Joyce, Nette and Isabel :) */

	return ret;
}

OPTIONS options_default(const OPTIONS* const opt1)
{
	assert(opt1);
	OPTIONS ret = *opt1;

	/* nthreads == 0 means default number */
	if (ret.nthread == 0) {
		/* leave some cores over for the operating system */		
		int n = sysconf(_SC_NPROCESSORS_ONLN) - 4;
		if (n < 0)
			n = 0;
		ret.nthread = n;
	}

	ret.estimate = estimconfig_default(&ret.estimate);
	ret.simulate = simconfig_default(&ret.simulate);

	return ret;
}

POPPARAM popparam_init(const POPMODEL* const popmodel,
					   const ADVANFUNCS* const advanfuncs,
					   const double eta[static OPENPMX_OMEGA_MAX])
						{
	return (POPPARAM) {
		.theta = popmodel->theta,
		.ntheta = popmodel->ntheta,
		.eta = eta,
		.nomega = popmodel->nomega,
		.sigma = popmodel->sigma,
		.nsigma = popmodel->nsigma,
		.nstate = advanfuncs->nstate,
	};
}

OPTIONS options_init_from_pmx(const OPENPMX* const pmx)
{
	var ret = (OPTIONS) {
		.nthread = pmx->nthread,
		._offset1 = pmx->_offset1,

		.verbose = pmx->verbose,
		.brief = pmx->brief,
	};
	return options_default(&ret);
}

static PMXSTATE* pmxstate_alloc(const OPENPMX* const pmx)
{
	let advanfuncs = advanfuncs_alloc(&pmx->data, &pmx->advan);
	let popmodel = popmodel_init(pmx->theta, pmx->omega, pmx->sigma);

	var temp = (PMXSTATE) {
		.advanfuncs = advanfuncs,
		.idata = idata_construct(&advanfuncs->recordinfo,
								 popmodel.ntheta,
								 popmodel.nomega,
								 popmodel.nsigma,
								 advanfuncs->nstate,
								 advanfuncs->advanconfig->imodelfields.size,
								 advanfuncs->advanconfig->predictfields.size),
		.tablecount = 0,
		.rng = 0,
	};

	/* have to allocate because its a opaque pointer so we have to memcpy */
	var ret = mallocvar(PMXSTATE, 1);
	memcpy(ret, &temp, sizeof(temp));
	return ret;
}

void pmxstate_ensure(OPENPMX* const pmx)
{
	if (pmx->state == 0)
		pmx->state = pmxstate_alloc(pmx);
}

static void pmxstate_free(PMXSTATE* pstate)
{
	advanfuncs_free((void*)pstate->advanfuncs); /* keeping advanfuncs const in PMXSTATE is nice, but this cludge is needed to free */
	idata_destruct(&pstate->idata);

	if (pstate->rng)
		gsl_rng_free(pstate->rng);

	scatter_cleanup();
	free(pstate);
}

OPENPMX pmx_copy(const OPENPMX* const pmx)
{
	OPENPMX ret = *pmx;
	ret.state = 0;
	return ret;
}

void pmx_copy_popparams(OPENPMX* dest, const OPENPMX* const src)
{
	memcpy(dest->theta, src->theta, sizeof(dest->theta));
	memcpy(dest->sigma, src->sigma, sizeof(dest->sigma));
	memcpy(dest->omega, src->omega, sizeof(dest->omega));

	dest->result = (PMXRESULT) { .objfn = DBL_MAX,
								 .type = OBJFN_INVALID,
								 .nfunc = 0 };
}

void pmx_cleanup(OPENPMX* pmx)
{
	pmxstate_free(pmx->state);
	pmx->state = 0;
}


