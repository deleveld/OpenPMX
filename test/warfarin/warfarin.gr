$DATA("models/warfarin_rate0.csv")
$ADVAN(diffeqn_libgsl)
	.nstate = 3,
$IMODEL(V, CL, KA, KOUT, E0, EMAX, C50)
	const double size = wt / 70.;
	const double vscale = pow(size, 1);
	const double cscale = pow(size, 0.75);
	const double kscale = pow(size, -0.25);
	V = THETA(1) * exp(ETA(1)) * vscale;
	CL = THETA(2) * exp(ETA(2)) * cscale;
	KA = THETA(3) * exp(ETA(3)) * kscale;
	E0 = THETA(5) + ETA(5);
	A_0(3, E0);
	EMAX = THETA(6);
	C50 = THETA(7) * exp(ETA(6));
	KOUT = exp(THETA(8));
$DIFFEQN
	DADT(1) = -A(1)*KA;
	DADT(2) = A(1)*KA - A(2)*(CL/V);
	double dcp = A(2)/V;
	double deff = dcp/(C50+dcp);
	double pd = 1 - EMAX*deff;
	DADT(3) = E0*KOUT*pd - KOUT*A(3);
$PREDICT(IPRED, BPRED)
	IPRED = A(2)/V;
	BPRED = A(3);
	if (dvid == 1) Y = IPRED*(1+ERR(1)) + ERR(2);
	if (dvid == 2) Y = BPRED + ERR(3);
$THETA
	{ 1,   8,   20, ESTIMATE },
	{ 0,   0.1,   1, ESTIMATE },
	{ 0.01,   0.5,   5, ESTIMATE },
	{ 0, 0, 0, FIXED },
	{ 50,   100,   150, ESTIMATE },	// E0
	{ 0,   1,   5, ESTIMATE },	// EMAX
	{ 0,   2,   5, ESTIMATE },	// C50
	{ -5,   -3,   1, ESTIMATE },	// log(KOUT)
$OMEGA(0.1, 0.1, 0.1, 0, 0, 0)
$SIGMA(1, 1, 100)
$MAIN
	estimate();
