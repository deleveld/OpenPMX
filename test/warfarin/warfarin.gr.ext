    ITERATION                   THETA1                   THETA2                   THETA3                   THETA4                   THETA5                   THETA6                   THETA7                   THETA8               SIGMA(1,1)               SIGMA(2,2)               SIGMA(3,3)               OMEGA(1,1)               OMEGA(2,1)               OMEGA(2,2)               OMEGA(3,1)               OMEGA(3,2)               OMEGA(3,3)               OMEGA(4,1)               OMEGA(4,2)               OMEGA(4,3)               OMEGA(4,4)               OMEGA(5,1)               OMEGA(5,2)               OMEGA(5,3)               OMEGA(5,4)               OMEGA(5,5)               OMEGA(6,1)               OMEGA(6,2)               OMEGA(6,3)               OMEGA(6,4)               OMEGA(6,5)               OMEGA(6,6)                      OBJ      RUNTIME       INEVAL
