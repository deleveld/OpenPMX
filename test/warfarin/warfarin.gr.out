$OUTFILE
OpenPMX 0.0.1 6b8fe54
config pthread 8
data records 526 used 526 removed 0
data individuals 32 observations 483
data table offset true
advan model GNU Scientific Library ODE
advan stepper gsl_odeiv2_step_rk8pd
advan nstate 3
advan abstol 1e-09 reltol 1e-09 hstart 1
optim FOCEI BOBYQA
checkout begin
tpool: start 0 1 2 3 4 5 6 7
warning: DV zero for observation: ID 8.000000 time 0.500000 record 2
error: AMT not finite (nan): ID 0.000000 time 0.000000 record 1
warning: DV zero for observation: ID 14.000000 time 0.500000 record 2
